segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -239,101 +259,183 @@ public int getMaxAppendLength() {
+     * @param toExecute An ordered List of Writes to execute.
      */
-    private void tryAppend(ArrayView data, WriteLedger writeLedger, CompletableFuture<LogAddress> result) {
-        long traceId = LoggerHelpers.traceEnterWithContext(log, this.traceObjectId, "tryAppend", data.getLength(), writeLedger.ledger.getId());
-        AsyncCallback.AddCallback callback = (rc, handle, entryId, ctx) -> {
-            @SuppressWarnings("unchecked")
-            CompletableFuture<LogAddress> completionFuture = (CompletableFuture<LogAddress>) ctx;
+    private boolean processWrites(List<Write> toExecute) {
+        for (int i = 0; i < toExecute.size(); i++) {
+            Write w = toExecute.get(i);
             try {
-                assert handle.getId() == writeLedger.ledger.getId() : "LedgerHandle.Id mismatch. Expected " + writeLedger.ledger.getId() + ", actual " + handle.getId();
-                if (rc != 0) {
-                    if (rc == BKException.Code.LedgerClosedException) {
-                        // LedgerClosed can happen because we just rolled over the ledgers. The way to detect this is to
-                        // check if the ledger we were trying to write to is closed and whether we have a new ledger.
-                        // If that's the case, retry the write with the new ledger.
-                        WriteLedger currentLedger = getWriteLedger();
-                        if (!currentLedger.ledger.isClosed() && currentLedger.ledger.getId() != writeLedger.ledger.getId()) {
-                            tryAppend(data, currentLedger, result);
-                            return;
-                        }
-                    }
-
-                    handleWriteException(rc, completionFuture);
-                    return;
+                // Record the beginning of a new attempt.
+                int attemptCount = w.beginAttempt();
+                if (attemptCount > this.config.getMaxWriteAttempts()) {
+                    // Retried too many times.
+                    throw new RetriesExhaustedException(w.getFailureCause());
                 }
 
-                // Successful write. Complete the callback future and update metrics.
-                LedgerAddress address = new LedgerAddress(writeLedger.metadata, entryId);
-                this.lastAppendAddress.set(address);
-                completionFuture.complete(address);
+                // Invoke the BookKeeper write.
+                w.getLedger().asyncAddEntry(w.data.array(), w.data.arrayOffset(), w.data.getLength(), this::addCallback, w);
             } catch (Throwable ex) {
-                completionFuture.completeExceptionally(ex);
+                // Synchronous failure (or RetriesExhausted). Fail current write.
+                boolean isFinal = !isRetryable(ex);
+                w.fail(ex, isFinal);
+
+                // And fail all remaining writes as well.
+                for (int j = i + 1; j < toExecute.size(); j++) {
```

Comment:
That's exactly what it does :)
`Write.fail` only finalizes the Write (with a failure) if the second argument (the boolean) is set to true. In this call, I set the second argument to `isFinal` - which is true only when the exception is not retryable. Otherwise the Write is marked and its attempt count preserved (it's already incremented). When we attempt to process it again, we see if we exceeded the max attempt count in which case we throw `RetriesExhaustedException`.